overlord/snapstate/snapstate_test.go

```diff
@@ -4228,76 +4229,72 @@ run-hook: Hold`)
 func (s *snapmgrTestSuite) verifyRefreshLast(c *C) {
 	var lastRefresh time.Time
 
-	tr := config.NewTransaction(s.state)
-	tr.Get("core", "refresh.last", &lastRefresh)
+	s.state.Get("last-refresh", &lastRefresh)
 	c.Check(time.Now().Year(), Equals, lastRefresh.Year())
 }
 
+func makeTestRefreshConfig(st *state.State) {
+	now := time.Now()
+	st.Set("last-refresh", time.Date(2009, 8, 13, 8, 0, 5, 0, now.Location()))
+
+	tr := config.NewTransaction(st)
+	tr.Set("core", "refresh.schedule", fmt.Sprintf("00:00-23:59"))
+	tr.Commit()
+}
+
 func (s *snapmgrTestSuite) TestEnsureRefreshesNoUpdate(c *C) {
 	s.state.Lock()
 	defer s.state.Unlock()
 	snapstate.CanAutoRefresh = func(*state.State) (bool, error) { return true, nil }
 
-	tr := config.NewTransaction(s.state)
-	tr.Set("core", "refresh.last", time.Time{})
-	tr.Commit()
+	makeTestRefreshConfig(s.state)
 
 	// Ensure() also runs ensureRefreshes()
 	s.state.Unlock()
 	s.snapmgr.Ensure()
+	// give the timer a bit of time
+	time.Sleep(100 * time.Millisecond)
 	s.state.Lock()
 
-	// nothing needs to be done, but refresh.last got updated
+	// nothing needs to be done, but last-refresh got updated
 	c.Check(s.state.Changes(), HasLen, 0)
 	s.verifyRefreshLast(c)
 }
 
-func (s *snapmgrTestSuite) TestEnsureRefreshesWithUpdate(c *C) {
+func (s *snapmgrTestSuite) TestEnsureRefreshesAlreadyRanInThisInterval(c *C) {
 	s.state.Lock()
 	defer s.state.Unlock()
 	snapstate.CanAutoRefresh = func(*state.State) (bool, error) { return true, nil }
 
+	now := time.Now()
+	fakeLastRefresh := now.Add(-1 * time.Hour)
+	s.state.Set("last-refresh", fakeLastRefresh)
+
 	tr := config.NewTransaction(s.state)
-	tr.Set("core", "refresh.last", time.Time{})
+	tr.Set("core", "refresh.schedule", fmt.Sprintf("00:00:00-%02d:%02d:%02d", now.Hour(), now.Minute(), now.Second()))
 	tr.Commit()
 
-	snapstate.Set(s.state, "some-snap", &snapstate.SnapState{
-		Active: true,
-		Sequence: []*snap.SideInfo{
-			{RealName: "some-snap", SnapID: "some-snap-id", Revision: snap.R(1)},
-		},
-		Current:  snap.R(1),
-		SnapType: "app",
-	})
-
-	// Ensure() also runs ensureRefreshes() and our test setup has an
-	// update for the "some-snap" in our fake store
+	// Ensure() also runs ensureRefreshes()
 	s.state.Unlock()
 	s.snapmgr.Ensure()
+	// give the timer a bit of time
+	time.Sleep(300 * time.Millisecond)
```

Comment:
Here it looks like it already created problems even on a high end system?

Perhaps same approach here and elsewhere if we have the same pattern.